### ISSUE TITLE
Deeper Optimization

### ISSUE THREAD

**Issue by FabianNitsche** [adletec/sonic#48]
Optimizing
`0.1 * var * 0.1`
does not lead to
`0.01 * var`
I created corresponding tests in https://github.com/FabianNitsche/sonic/tree/optimizationIssue

Do you have an idea to allow deeper level optimization? Or is this out of scope?

I started to implement something, but it is really ugly. See Optimizer.OptimizeNew

**Comment by vlow** [comment]
Interesting find!

At first glance, an easy solution could be to sort subsequent multiplications by type. `0.1 * 0.1 * var` should be correctly optimized.

However, this would probably add something in the order of O(n log(n)) to the complexity of the optimizer, and it still won't be sufficient to fix the other good findings you wrote tests for, e.g. `0,1 * (var * 0,1)`.

The AST of `0.1 * var * 0.1` might look like that:

```
                                                                                 
        var              0.1                                                     
          \               /                                                      
           \             /                                                       
            \           /                                                        
             \         /                                                         
              \       /                                                          
 0.1           -  *  -                                                           
  \               /                                                              
   \             /                                                               
    \           /                                                                
     \         /                                                                 
      \       /                                                                  
       -  *  -                                                                   

```
At the moment, the optimizer works on a per operation basis, i.e. it will recurse down the tree and check if the operation can be evaluated. So it'll check if `0.1 * multiplicationOperation` can be evaluated, then recurse into `multiplicationOperation` and check if `var * 0.1` can be evaluated (no) and return down the tree with the sad news.

This is pretty fast, so optimization is a fast default option. Deeper optimization might certainly be possible, but we might leave the happy path of O(n). This makes the decision to implement deeper optimization more complicated, since it might actually make _sonic_ considerably slower for many expressions which don't benefit from constant folding.

But I'm open to suggestions here. Maybe there is a way we can at least improve some cases without changing the time complexity of the optimizer?

**Comment by FabianNitsche** [comment]
I do not think that it will be such a performance hit, since if one restricts the deeper optimization to `*`.`/` and `+`,`-` it will rarly go down many levels. I think it is a rare case where that happens.
I was just curious if you know some kind of magic algorithm that does that efficiently. I was pondering my brain and just came up with the ugly solution for `*` that is in my branch...

**Comment by FabianNitsche** [comment]
Additionally, parsing and optimization is probaby not the performance critical thing, the execution is. Since you probably parse a formula once to calcuate thousands of values.

**Comment by vlow** [comment]
> I do not think that it will be such a performance hit, since if one restricts the deeper optimization to *./ and +,- it will rarly go down many levels. I think it is a rare case where that happens.

Could be, somewhat depends on the actual implementation. I'm not sure if I can assess the performance impact for all corner cases halfway reliably, but we have a pretty extensive benchmarking suite, so I'd prefer measuring over guessing anyway 😄.

> Additionally, parsing and optimization is probaby not the performance critical thing, the execution is. Since you probably parse a formula once to calcuate thousands of values.

That actually kind of depends on the usage. I also assume that re-evaluating expressions with different values might be the most common use case, but I have too little information on the actual usage of _sonic_ to rely on that. However, you're still right in that someone who wants to evaluate a lot of expressions only once doesn't benefit from the Optimizer in the first place and will be a lot faster disabling optimizer, caching and using interpreted evaluation.

But for very long expression strings, changing the time complexity of the Optimizer still might have a noticable impact, and we definitely have customers with those, so I would at least want to run a couple of real-world benchmarks before merging anything like that to main.

> I was just curious if you know some kind of magic algorithm that does that efficiently.

Unfortunately, not from the top of my head. But let me sleep on it, maybe inspiration strikes tomorrow ;)
